src/Features/Core/Portable/SignatureHelp/SignatureHelpService.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Extensions;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SignatureHelp;
+
+/// <summary>
+/// A service that is used to determine the appropriate signature help for a position in a document.
+/// </summary>
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+[Export(typeof(ISignatureHelpService)), Shared]
+internal sealed class SignatureHelpService([ImportMany] IEnumerable<Lazy<ISignatureHelpProvider, OrderableLanguageMetadata>> allProviders)
+    : ISignatureHelpService
+{
+    private readonly ConcurrentDictionary<string, ImmutableArray<ISignatureHelpProvider>> _providersByLanguage = [];
+    private readonly IEnumerable<Lazy<ISignatureHelpProvider, OrderableLanguageMetadata>> _allProviders = allProviders;
+
+    private ImmutableArray<ISignatureHelpProvider> GetProviders(string language)
+    {
+        return _providersByLanguage.GetOrAdd(language, language
+            => _allProviders.Where(p => p.Metadata.Language == language)
+                .SelectAsArray(p => p.Value));
```

Comment:
nit.  the wrapping of the => def confused me here.  Consider:

```suggestion
        return _providersByLanguage.GetOrAdd(language, language =>
            _allProviders
                .Where(p => p.Metadata.Language == language)
                .SelectAsArray(p => p.Value));
```

Or

```suggestion
        return _providersByLanguage.GetOrAdd(language, language =>
            _allProviders.Where(p => p.Metadata.Language == language).SelectAsArray(p => p.Value));
```

Also, i believe there's a SelectAsArray that already takes a predicate.  so you may be able to do:

```suggestion
        return _providersByLanguage.GetOrAdd(language, language =>
            _allProviders.SelectAsArray(p => p.Metadata.Language == language, p => p.Value));
```